nameserver/dns.go

```diff
@@ -1,117 +1,263 @@
 package nameserver
 
 import (
+	"bytes"
+	"fmt"
 	"math/rand"
+	"net"
+	"strings"
 	"time"
 
 	"github.com/miekg/dns"
+
+	. "github.com/weaveworks/weave/common"
+	"github.com/weaveworks/weave/net/address"
 )
 
 const (
-	minUDPSize = 512
-	maxUDPSize = 65535
+	topDomain        = "."
+	reverseDNSdomain = "in-addr.arpa."
+	etcResolvConf    = "/etc/resolv.conf"
+	udpBuffSize      = uint16(4096)
+	minUDPSize       = 512
 )
 
-func makeHeader(r *dns.Msg, q *dns.Question, ttl int) *dns.RR_Header {
-	return &dns.RR_Header{
-		Name:   q.Name,
-		Rrtype: q.Qtype,
-		Class:  dns.ClassINET,
-		Ttl:    uint32(ttl),
+type DNSServer struct {
+	ns     *Nameserver
+	domain string
+	ttl    uint32
+	port   int
+
+	servers   []*dns.Server
+	upstream  *dns.ClientConfig
+	tcpClient *dns.Client
+	udpClient *dns.Client
+}
+
+func NewDNSServer(ns *Nameserver, domain string, port int, ttl uint32,
+	clientTimeout time.Duration) (*DNSServer, error) {
+
+	s := &DNSServer{
+		ns:        ns,
+		domain:    domain,
```

Comment:
Added in https://github.com/weaveworks/weave/commit/1bbcd7bbf2922d95b9761b7f2f5756444a068801